Suggest Replacement Types for Invalid Placeholders

CodaFi opened the pull request:
First, relax the restrictions on placeholder types appearing in interface types. In the past, we would pretty much replace every instance of them with an error type, which would lose information when we came around to apply the contextual type to any solutions for associated inference sources like initializers in pattern bindings or the return types of functions. Preserve these types in interface types so they can be opened later.

Second, now that placeholder types are preserved, we can open them wherever they appear in positions where they are used as contextual types. Use the checks we already run in the primaries to ban placeholders in top-level positions. Only now, use the inferred type of any associated expressions or statements to present the user with a contextual replacement type.

For parameters in functions and enum cases, we use any default argument expressions to get the contextual type.

For functions, we use a traversal similar to the opaque result type finder to find the type of any return statements in the program and present those as options.

Resolves rdar://82837146

CodaFi committed:
Revert "[Sema] Diagnose and reject top-level placeholders"

This reverts commit f6b6bff6ded2c041dbed14b7c3cea45d43fd3c5a.

CodaFi committed:
Relax Restrictions on Placeholder Types Appearing in Interface Types

These restrictions are meant to keep placeholder types from escaping TypeCheckType. But there's really no harm in that happening as long as we diagnose it on the way out in the places it's banned. (We also need to make sure we're only diagnosing things in primaries, but that's a minor issue). The end result is that we lose information because a lot of the AST that has placeholders in it becomes filled with error types instead.

Lift the restriction on placeholders appearing in the interface type, teach the mangler to treat them as unresolved types, and teach serialization to treat them as error types.

CodaFi committed:
Suggest Replacement Types for Invalid Placeholders

Now that placeholder types are preserved, we can open them wherever they appear in positions where they are used as contextual types. Use the checks we already run in the primaries to ban placeholders in top-level positions. Only now, use the inferred type of any associated expressions or statements to present the user with a contextual return type.

For parameters in functions and enum cases, we use any default argument expressions to get the contextual type.

For functions, we use a traversal similar to the opaque result type finder to find the type of any return statements in the program and present those as options.

CodaFi commented:
@swift-ci test

Jumhyn commented on lib/Sema/TypeCheckDeclPrimary.cpp:
```diff
@@ -967,11 +967,67 @@ static void checkInheritedDefaultValueRestrictions(ParamDecl *PD) {
+      TypeChecker::notePlaceholderReplacementTypes(
+          ifacety, expr->getType()->mapTypeOutOfContext());
+    } else {
+      param->diagnose(diag::placeholder_type_not_allowed);
```
Is there a reason why we use a generic fallback message here? Seems like the same diagnostic we use above, just without the fix-it, would be appropriate here as well.

Jumhyn reviewed (commented):
This is awesome, major diagnostic improvement! I have a couple comments about diagnostics and some thoughts on allowing top-level placeholders wrt the evolution process.

xwu commented on test/Sema/placeholder_type.swift:
```diff
@@ -1,16 +1,17 @@
 // RUN: %target-typecheck-verify-swift
 
-let x: _ = 0 // expected-error {{placeholders are not allowed as top-level types}}
+let x: _ = 0
```
Without opening up the question of top-level use in general, where _not having a type annotation at all_ would compile as here, it’s not a load-bearing “use,” and perhaps could be considered not a “use” at all.

Jumhyn replied:
Perhaps this wasn't the best line to attach the comment to, but we have another instance below that veers into the more problematic territory that @xwu noted:

```
let _: [String: Int] = dictionary(ofType: _.self) // used to fail, compiles under this PR
```

IMO we should seek core team guidance (if you haven't already) before accepting a change that directly violates the accepted language of SE-0315 (even though I'd be in favor of the change, evolution concerns notwithstanding 🙂).

xedin commented on lib/AST/ASTContext.cpp:
```diff
@@ -3594,7 +3594,6 @@ GenericFunctionType *GenericFunctionType::get(GenericSignature sig,
                                               Optional<ExtInfo> info) {
   assert(sig && "no generic signature for generic function type?!");
   assert(!result->hasTypeVariable());
-  assert(!result->hasPlaceholder());
```
Could you add an assert that verifies that there are placeholder no placeholder that originated from a type variable directly contain one (e.g. dependent member types with type variable base)?

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3683,6 +3681,12 @@ NOTE(descriptive_generic_type_declared_here,none,
+ERROR(placeholder_type_not_allowed_in_return_type,none,
+      "type placeholder may not appear in function return type", ())
+ERROR(placeholder_type_not_allowed_in_parameter,none,
+      "type placeholder may not appear in top-level parameter", ())
```
I agree, we need to find something more user-friendly here.

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3683,6 +3681,12 @@ NOTE(descriptive_generic_type_declared_here,none,
+ERROR(placeholder_type_not_allowed_in_parameter,none,
+      "type placeholder may not appear in top-level parameter", ())
+NOTE(replace_placeholder_with_inferred_type,none,
+     "replace the placeholder with the correct type %0", (Type))
```
Both `the` are extraneous here according to Diagnostics.md and I'm not sure whether we should use `correct` at all...

CodaFi committed:
Use Parameter-Specific Diagnostic When Diagnosing Parameters

CodaFi committed:
[NFC] s/Correct Type/Inferred Type

CodaFi committed:
Restore Top-Level Placeholder Restrictions in Expression Position

Respect the text of https://github.com/apple/swift-evolution/blob/main/proposals/0315-placeholder-types.md#top-level-type-placeholders and re-restrict the usage of placeholders in top-level expressions.

CodaFi commented:
@swift-ci test

CodaFi commented:
@swift-ci test

CodaFi commented:
⛵
